src/services/bonusService.ts

```diff
@@ -114,10 +125,96 @@ export default class BonusService {
-            return ResponseErrorInternal(
-              "Underlying API fails with an unexpected 401"
+            return ResponseErrorUnexpectedAuthProblem();
+          default:
+            return ResponseErrorStatusNotDefinedInSpec(response);
+        }
+      });
+    });
+
+  /**
+   * Retrieve a bonus activation detail decorated with its qr-code images generated at runtime.
+   */
+  public readonly getLatestBonusActivationById = (
+    user: User,
+    bonusId: NonEmptyString
+  ): Promise<
+    // tslint:disable-next-line: max-union-size
+    | IResponseErrorInternal
+    | IResponseSuccessAccepted
+    | IResponseErrorNotFound
+    | IResponseSuccessJson<BonusActivation>
+  > =>
+    withCatchAsInternalError(async () => {
+      const validated = await this.bonusApiClient.getLatestBonusActivationById({
+        bonus_id: bonusId,
+        fiscalCode: user.fiscal_code
+      });
+
+      // tslint:disable-next-line: no-identical-functions
+      return withValidatedOrInternalError(validated, response => {
+        switch (response.status) {
+          case 200:
+            // TODO: aggiungere qr code
+            return ResponseSuccessJson(response.value);
+          case 404:
+            return ResponseErrorNotFound(
+              "BonusActivation not found",
+              `Could not find a bonus activation for fiscal code: ${user.fiscal_code} and bonus id ${bonusId}`
             );
+          case 500:
+            return ResponseErrorInternal(readableProblem(response.value));
+          case 401:
+            return ResponseErrorUnexpectedAuthProblem();
+          default:
+            return ResponseErrorStatusNotDefinedInSpec(response);
+        }
+      });
+    });
+
+  /**
+   * Start bonus activation request procedure
+   */
+  public readonly startBonusActivationProcedure = (
+    user: User
+  ): Promise<
+    // tslint:disable-next-line: max-union-size
+    | IResponseErrorInternal
+    | IResponseErrorConflict
+    | IResponseErrorValidation
+    | IResponseSuccessAccepted
+    | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
+  > =>
+    withCatchAsInternalError(async () => {
+      const validated = await this.bonusApiClient.startBonusActivationProcedure(
+        {
+          fiscalCode: user.fiscal_code
+        }
+      );
+
+      return withValidatedOrInternalError(validated, response => {
+        switch (response.status) {
+          case 201:
+            return ResponseSuccessRedirectToResource(
+              response.value,
+              response.headers.Location || "",
+              response.value
+            );
+          case 202:
+            return ResponseSuccessAccepted();
+          case 409:
+            return ResponseErrorConflict(
+              "Cannot activate a new bonus because another active or consumed bonus related to this user was found."
+            );
+          case 403:
+            return ResponseErrorValidation(
```

Comment:
why validation?

Reply:
I think it's actually a validation failure (you don't have prerequisites for the operation).

Reply:
we should forward the response status (403) in this case to the app trying to keep the specs betweebn functions and be aligned